File: src/lib/encrypter.c
#include <why2/encrypter.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <sys/time.h>
#include <unistd.h>

#include <why2/flags.h>
#include <why2/misc.h>

outputFlags encryptText(char *text, char *keyNew, inputFlags flags)
{
    //TIME VARIABLES
    struct timeval startTime;
    struct timeval finishTime;
    gettimeofday(&startTime, NULL);

    //CHECK FOR ACTIVE VERSION
    if (!flags.skipCheck) checkVersion(flags);

    //CHECK FOR INVALID text
    checkText(text, flags);

    if (strcmp(text, "") == 0)
    {
        if (!flags.noOutput) fprintf(stderr, "No text to encrypt!\n");
        exit(1);
    }

    //VARIABLES
    char *key = malloc(getKeyLength());
    char *returningText;
    char *textBuffer;
    int *textKeyChain = malloc(sizeof(int) * strlen(text));
    int numberBuffer;
    FILE *fileBuffer;

    //TRY TO MAKE RANDOM GENERATION REALLY "RANDOM"
    fileBuffer = fopen("/dev/urandom", "r");
    fread(&numberBuffer, 4, 1, fileBuffer);
    srand(numberBuffer);
    fclose(fileBuffer);

    if (keyNew != NULL)
    {
        checkKey(keyNew, flags); //CHECK FOR INVALID key

        strcpy(key, keyNew);

        //REDEFINE keyLength
        setKeyLength(strlen(key));

        goto skipKey;
    }

    //LOAD KEY
    for (int i = 0; i < getKeyLength(); i++)
    {
        //SET numberBuffer TO RANDOM NUMBER BETWEEN 0 AND 52
        numberBuffer = (rand() % 52) + 1;

        //GET CHAR FROM numberBuffer
        if (numberBuffer > 26)
        {
            numberBuffer += 70;
        } else
        {
            numberBuffer += 64;
        }

        key[i] = (char) numberBuffer;
    }

    key[getKeyLength()] = '\0'; //TODO: CHECK IF THIS CAN'T CAUSE SOME LITTLE TROUBLES (SHOULDN'T)

    skipKey:

    //LOAD textKeyChain
    generateTextKeyChain(key, textKeyChain, strlen(text));

    //ACTUALLY ENCRYPT TEXT
    for (int i = 0; i < strlen(text); i++)
    {
        textKeyChain[i] -= (int) text[i];
    }

    //COUNT REQUIRED SIZE FOR returningText
    for (int i = 0; i < strlen(text); i++)
    {
        numberBuffer += countIntLength(textKeyChain[i]);
    }

    //ALLOCATE returningText (WITH THE SEPARATORS)
    returningText = malloc(numberBuffer + strlen(text) - 1);
    strcpy(returningText, "");

    //LOAD returningText
    for (int i = 0; i < strlen(text); i++)
    {
        numberBuffer = sizeof(int) * countIntLength(textKeyChain[i]);

        if (i != 0)
        {
            textBuffer = realloc(textBuffer, numberBuffer);
        } else
        {
            textBuffer = malloc(numberBuffer);
        }

        sprintf(textBuffer, "%d", textKeyChain[i]);

        strcat(returningText, textBuffer);

        if (i != strlen(text) - 1)
        {
            strcat(returningText, ENCRYPTION_SEPARATOR_STRING);
        }
    }

    //GET FINISH TIME
    gettimeofday(&finishTime, NULL);

    //LOAD output
    outputFlags output =
    {
        returningText, //ENCRYPTED TEXT
        key, //GENERATED/USED KEY
        countUnusedKeySize(text, key), // NUMBER OF UNUSED CHARS IN KEY
        compareTimeMicro(startTime, finishTime) // ELAPSED TIME
    };

    //DEALLOCATION
    free(textKeyChain);
    free(textBuffer);

    return output;
}

File: src/encrypter.js
module.exports = async (text, key = undefined) => {
	const KEY_LENGTH = 50
        const ENCRYPTION_SEPARATOR_STRING = '.';
        const GENERATE_RANDOM_KEY = async (numberBuffer) => {
            key = [];
            //LOAD KEY
            for (var i = 0; i < KEY_LENGTH; i++) {
                //SET numberBuffer TO RANDOM NUMBER BETWEEN 0 AND 52
                numberBuffer = await crypto.randomInt(0, 52 + 1);

                //GET CHAR FROM numberBuffer
                if (numberBuffer > 26) {    
                    numberBuffer += 70;
                } else {
                    numberBuffer += 64;
                }

                key[i] = numberBuffer;
            }
        }

        //VARIABLES
        var keyRand = KEY_LENGTH;
        var returningText = new String();
        var textBuffer;
        var textKeyChain = [];
        textKeyChain.length = text.length;
        var numberBuffer;

        if (key < KEY_LENGTH) {
            return { exitCode: -2 };
        }
        if (!key) {
            await GENERATE_RANDOM_KEY(numberBuffer);
        }
        //LOAD textKeyChain
        for (var i = 0; i < textKeyChain.length; i++)  {
            numberBuffer = i;

            //CHECK, IF numberBuffer ISN'T GREATER THAN KEY_LENGTH AND CUT UNUSED LENGTH
            while (numberBuffer >= key.length)
            {
                numberBuffer -= key.length;
            }

            if (typeof key === 'string') {
                key = key.split("");
                key = key.map(element => element.charCodeAt(0));
            }

            //FILL textKeyChain
            if ((numberBuffer + 1) % 3 == 0)
            {
                textKeyChain[i] = key[numberBuffer] * key[numberBuffer + 1];
            } else if ((numberBuffer + 1) % 2 == 0)
            {
                textKeyChain[i] = key[numberBuffer] - key[numberBuffer + 1];
            } else
            {
                textKeyChain[i] = key[numberBuffer] + key[numberBuffer + 1];
            }
        }

        //ACTUALLY ENCRYPT TEXT
        for (var i = 0; i < text.length; i++) {
            textKeyChain[i] -= text[i].charCodeAt(0);
        }

        numberBuffer = 0;

        //LOAD returningText
        for (var i = 0; i < textKeyChain.length; i++) {
            textBuffer = Math.floor(Math.log10(Math.abs(textKeyChain[i])));

            textBuffer = textKeyChain[i].toString();

            returningText += textBuffer;

            if (i != textKeyChain.length - 1)
            {
                returningText += ENCRYPTION_SEPARATOR_STRING;
            }

            textBuffer = undefined;
        }
        key = key.map(element => String.fromCharCode(element)).join("");
        return { exitCode: 0, value: returningText, key: key };
}